In [1]:
import numpy as np
import gudhi
import mdtraj as md
import seaborn as sns
import itertools
import matplotlib as mpl
import matplotlib.pyplot as plt

In [2]:
traj = md.load_dcd('deca56D_prd.dcd', top='deca56D_clean.pdb')
traj = md.Trajectory.superpose(traj, traj[0], frame=0, atom_indices=None, ref_atom_indices=None, parallel=True)

In [343]:
frame = traj[10000]
frame.save("10000.pdb")

In [3]:
carbon_atoms = traj.topology.select("element == 'C'")

In [230]:
frame3 = traj.xyz[0, carbon_atoms, :]

In [242]:
frame3 = traj.xyz[500, carbon_atoms, :]

In [252]:
frame3 = traj.xyz[3000, carbon_atoms, :]

In [262]:
frame3 = traj.xyz[10000, carbon_atoms, :]

In [271]:
frame3 = traj.xyz[60000, carbon_atoms, :]

In [280]:
frame3 = traj.xyz[80000, carbon_atoms, :]

In [289]:
frame3 = traj.xyz[160000, carbon_atoms, :]

In [298]:
frame3 = traj.xyz[240000, carbon_atoms, :]

In [307]:
frame3 = traj.xyz[380000, carbon_atoms, :]

In [316]:
frame3 = traj.xyz[550000, carbon_atoms, :]

In [325]:
frame3 = traj.xyz[620000, carbon_atoms, :]

In [253]:
num_rows = np.size(frame3, axis=0)
print(f"Number of rows: {num_rows}")

Number of rows: 60


In [326]:
rips_complex = gudhi.RipsComplex(points=frame3)

In [327]:
simplex_tree = rips_complex.create_simplex_tree(max_dimension=3)

In [328]:
persistence = simplex_tree.persistence()

In [329]:
alpha_complex = {tuple(sorted(simplex)) : radius for simplex, radius in simplex_tree.get_filtration()}
boundary_matrix = {simplex : set(itertools.combinations(simplex, len(simplex)-1))-{tuple()} for simplex in alpha_complex}

R = { k : v for k,v in boundary_matrix.items()}
V = { k : {k} for k in boundary_matrix}
lowinv = {} # lowinv[i]=index of column with the lowest 1 at i
order_function = lambda s: (alpha_complex[s], s)
for s in sorted(R, key=order_function):
    t = lowinv.get(max(R[s], key=order_function),-1) if len(R[s])!=0 else -1
    while t!=-1:
        R[s] = R[t]^R[s] # symmetric difference of t-th and s-th columns
        V[s] = V[t]^V[s]
        t = lowinv.get(max(R[s], key=order_function),-1) if len(R[s])!=0 else -1
    if len(R[s])!=0:
        lowinv[max(R[s], key=order_function)] = s

loops = [bar for bar in lowinv.items() if len(bar[0])==2]
longest_loop = max(loops, key=lambda bar: alpha_complex[bar[1]]-alpha_complex[bar[0]])
representative_1 = R[longest_loop[1]]
representative_2 = V[longest_loop[0]]

In [330]:
print(representative_1)
#print(representative_2)

{(36, 39), (31, 32), (13, 19), (6, 7), (24, 25), (19, 24), (13, 14), (13, 17), (38, 39), (6, 11), (14, 17), (32, 33), (7, 13), (25, 31), (36, 38), (11, 33)}


In [350]:
carbon_atoms[46]

146

list = []

In [332]:
list.append(representative_1)

In [333]:
list

[{(25, 26),
  (25, 30),
  (26, 27),
  (27, 46),
  (30, 36),
  (36, 37),
  (37, 42),
  (42, 46)},
 {(42, 43), (42, 47), (43, 49), (47, 58), (49, 57), (57, 58)},
 {(25, 27),
  (25, 30),
  (27, 46),
  (30, 31),
  (31, 37),
  (37, 42),
  (42, 45),
  (45, 46)},
 {(24, 25), (24, 28), (25, 31), (28, 46), (31, 37), (37, 42), (42, 46)},
 {(0, 1),
  (0, 2),
  (1, 2),
  (1, 6),
  (1, 7),
  (6, 9),
  (7, 12),
  (9, 10),
  (10, 57),
  (12, 13),
  (13, 19),
  (19, 24),
  (24, 25),
  (24, 26),
  (24, 28),
  (25, 31),
  (26, 28),
  (31, 36),
  (36, 37),
  (37, 42),
  (42, 43),
  (42, 45),
  (42, 47),
  (43, 48),
  (45, 46),
  (46, 47),
  (48, 49),
  (49, 55),
  (54, 55),
  (54, 56),
  (56, 57)},
 {(42, 43), (42, 47), (43, 49), (47, 58), (49, 54), (54, 57), (57, 58)},
 {(25, 27),
  (25, 30),
  (27, 46),
  (30, 31),
  (31, 37),
  (37, 42),
  (42, 45),
  (45, 46)},
 {(25, 27), (25, 31), (27, 45), (31, 37), (37, 42), (42, 45)},
 {(13, 17),
  (13, 19),
  (17, 47),
  (19, 24),
  (24, 25),
  (25, 30),
  (30,

#traj[1642600] b-hairpin
#index 139 or index 169 or index 29 or index 143 or index 141 or index 137 or index 118 or index 120 or index 131 or index 23 or index 25 or index 48 or index 51

#traj[760000] b-hairpin
#index 99 or index 105 or index 82 or index 80 or index 63 or index 61 or index 44 or index 42 or index 36 or index 25

#traj[1300000] b-hairpin
#index 44 or index 46 or index 48 or index 63 or index 80 or index 82 or index 101 or index 118 or index 122 or index 131 or index 165

#traj[1642623] b-hairpin
#index 23 or index 25 or index 27 or index 29 or index 44 or index 46 or index 51 or index 99 or index 101 or index 103 or index 105 or index 118 or index 120 or index 137 or index 139 or index 156 or index 160 or index 169

#traj[500] unfolded
#index 137 or index 139 or index 150 or index 158 or index 182 or index 185

#traj[10000] unfolded
#index 80 or index 82 or index 90 or index 101 or index 120 or index 137 or index 146